vowpalwabbit/io_buf.cc

```diff
@@ -158,3 +158,26 @@ void io_buf::close_file_or_socket(int f)
   close(f);
 #endif
 }
+
+//make all data that was writen available for reading
+void io_buf::flip_from_write_to_read()
+{
+  space.end() = head;
+  head = space.begin();
+}
+
+size_t io_buf::copy_to(void *dst, size_t max_size)
+{
+  size_t to_copy = std::min(available_to_read(), max_size);
+  memcpy(dst, space.begin(), to_copy);
+  return to_copy;
+}
+
+void io_buf::replace_buffer(char *buff, size_t capacity)
+{
+  // TODO the following should be moved to v_array
+  space.delete_v();
+  space.begin() = buff;
+  space.end_array = space.end() = buff + capacity;
```

Comment:
space.end() should be set to space.begin(), right? Or is the assumtion that the replacement buffer is filled to capacity when this function is called?

Reply:
No, `end() == begin()` only makes sense for writing (actually, end is never used).
When reading, end() must point one after the last readable byte.

Reply:
Ok, so the assumption is that the replacement buffer is filled with data. Can you add a comment making that clear?